### PR TITLE
Issue#2 add cache

### PR DESCRIPTION
**Pull request by ilayzen**

### CODE REVIEW

File: pkg/plugin/oauth2/setup.go [hunk 1]
```diff
@@ -152,7 +158,42 @@ func setupOAuth2(def *proxy.RouterDefinition, rawConfig plugin.Config) error {
...
+	cache := inMemory.NewCache()
+
+	url := fmt.Sprintf("%s/roles", kafkaConfig.RBACUrl)
+
+	resp, err := http.DefaultClient.Get(url)
+	if err != nil {
+		return err
+	}
+
+	body, err := io.ReadAll(resp.Body)
```
Comment (taaraora) [inline pkg/plugin/oauth2/setup.go]: body.Close()

File: pkg/plugin/oauth2/setup.go [hunk 2]
```diff
@@ -152,7 +158,42 @@ func setupOAuth2(def *proxy.RouterDefinition, rawConfig plugin.Config) error {
 		return err
 	}
 
-	def.AddMiddleware(NewKeyExistsMiddleware(manager))
+	var kafkaConfig config.Config
+
+	_ = config.UnmarshalYAML("./config/config.yaml", &kafkaConfig)
+	if kafkaConfig.KafkaAddr == "" {
+		fmt.Errorf("Kafka adress not setted!")
```
Comment (taaraora) [inline pkg/plugin/oauth2/setup.go]: is it fatal?

File: pkg/plugin/oauth2/middleware.go [hunk 3]
```diff
@@ -94,3 +114,27 @@ func NewKeyExistsMiddleware(manager Manager) func(http.Handler) http.Handler {
...
+						for _, endpoint := range feature.Endpoints {
+							if endpoint.Method == method && endpoint.Path == path {
+								return nil
+							}
+						}
+					}
+				}
+			}
+		}
+	}
```
Comment (taaraora) [inline pkg/plugin/oauth2/middleware.go]:
```
	for _, userRole := range rolesOfUser {
		for _, role := range cache.Roles {
			if role.Name != userRole {
				continue
			}

			for _, feature := range role.Features {
				for _, endpoint := range feature.Endpoints {
					if endpoint.Method == method && endpoint.Path == path {
						return nil
					}
				}
			}
		}
	}
	```

File: pkg/plugin/oauth2/middleware.go [hunk 4]
```diff
@@ -71,6 +73,24 @@ func NewKeyExistsMiddleware(manager Manager) func(http.Handler) http.Handler {
...
+				return
+			}
+
+			if len(claims.Roles) <= 0 {
+				errors.Handler(w, r, fmt.Errorf("No roles have been set"))
+				return
+			}
+
+			err = RolesChecker(cache, claims.Roles, r.URL.Path, r.Method)
+			if err != nil {
```
Comment (taaraora) [inline pkg/plugin/oauth2/middleware.go]: please handle errors properly

**Review by taaraora — CHANGES_REQUESTED** [review]

File: pkg/plugin/oauth2/setup.go (hunk 1 above)
Comment (ilayzen) [reply pkg/plugin/oauth2/setup.go]: fixed

Comment (ilayzen) [reply pkg/plugin/oauth2/setup.go:178]: fixed

File: pkg/plugin/oauth2/setup.go (hunk 2 above)
Comment (ilayzen) [reply pkg/plugin/oauth2/setup.go]: this is a strange error, there is a config file in the docker container, i can read it but it always drops here.

File: pkg/cache/roles.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	return nil
+}
+
+func (c *RolesCache) Get(roleName string) (*models.Role, error) {
+	c.RLock()
+	defer c.RUnlock()
+
+	role, found := c.Roles[roleName]
+	if !found {
+		return nil, fmt.Errorf("Cant get role")
```
Comment (ribaraka) [inline pkg/cache/roles.go]: pls add `roleName` var to err message

File: pkg/cache/roles.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+	return role, nil
+}
+
+func (c *RolesCache) Delete(roleName string) error {
+	c.Lock()
+	defer c.Unlock()
+
+	if _, found := c.Roles[roleName]; !found {
+		log.Println("Key not found")
```
Comment (ribaraka) [inline pkg/cache/roles.go]: return nil, fmt.Errorf("Cannot get a %s role", roleName)

**Review by ribaraka — CHANGES_REQUESTED** [review]

**Pull request closed without merging**